File: tools/parsing-question/parser/content_extractor.py
"""
Content Extractor

Implements the 7-step content cleaning process for LaTeX questions.
"""

import re
from typing import List, Tuple
import sys
import os

# Add parent directory to path for imports
current_dir = os.path.dirname(os.path.abspath(__file__))
parent_dir = os.path.dirname(current_dir)
sys.path.insert(0, parent_dir)

from parser.bracket_parser import BracketParser


class ContentExtractor:
    """
    Extracts and cleans question content from LaTeX following the 7-step process:
    1. Extract content from ex environment
    2. Remove metadata patterns
    3. Handle layout commands
    4. Remove image commands
    5. Remove answer commands
    6. Remove solution section
    7. Normalize whitespace
    """
    
    # Patterns for metadata removal
    METADATA_PATTERNS = [
        r'%\s*\[\s*[0-9A-Z]{5,6}(?:-[0-9A-Z])?\s*\]\s*%?',  # QuestionCode
        r'%\s*\[\s*Nguồn:?[^\]]+\s*\]\s*%?',                 # Source
        r'\[\s*[A-Z]{2}\.\d+\s*\]',                          # Subcount
    ]
    
    # Patterns for image environments
    IMAGE_PATTERNS = [
        r'\\begin\{center\}[\s\S]*?\\end\{center\}',
        r'\\begin\{tikzpicture\}[\s\S]*?\\end\{tikzpicture\}',
        r'\\includegraphics(?:\[[^\]]*\])?\{[^}]*\}',
        r'\\begin\{figure\}[\s\S]*?\\end\{figure\}',
    ]
    
    # Patterns for answer commands
    ANSWER_PATTERNS = [
        r'\\choice(?:\[[0-9]\])?(?:\s*\{[^}]*\})*',
        r'\\choiceTF(?:\[[t12]\])?(?:\s*\{[^}]*\})*',
        r'\\choiceTFt(?:\s*\{[^}]*\})*',
        r'\\shortans(?:\[[a-z0-9]+\])?(?:\s*\{[^}]*\})*',
        r'\\matching(?:\s*\{[^}]*\})*',
    ]
    
    # Pattern for solution section
    SOLUTION_PATTERN = r'\\loigiai\s*\{[^}]*(?:\{[^}]*\}[^}]*)*\}'
    
    @classmethod
    def extract_from_ex_environment(cls, latex_content: str) -> str:
        """
        Step 1: Extract content from \begin{ex}...\end{ex} environment.
        
        Args:
            latex_content: Full LaTeX content
            
        Returns:
            Content inside ex environment
        """
        environments = BracketParser.extract_environment_content(latex_content, "ex")
        if environments:
            return environments[0]  # Return first ex environment
        return ""
    
    @classmethod
    def remove_metadata_patterns(cls, content: str) -> str:
        """
        Step 2: Remove metadata patterns (questionCode, source, subcount).
        
        Args:
            content: Content to clean
            
        Returns:
            Content with metadata removed
        """
        for pattern in cls.METADATA_PATTERNS:
            content = re.sub(pattern, '', content, flags=re.MULTILINE)
        return content
    
    @classmethod
    def handle_layout_commands(cls, content: str) -> str:
        """
        Step 3: Handle layout commands (immini).
        
        For 2-column layout with \immini[thm]{content1}{content2}, keep content1 only.
        
        Args:
            content: Content to process
            
        Returns:
            Content with layout commands processed
        """
        # Pattern for \immini[optional]{content1}{content2}
        immini_pattern = r'\\immini(?:\[[^\]]*\])?\s*\{([^}]*(?:\{[^}]*\}[^}]*)*)\}\s*\{[^}]*(?:\{[^}]*\}[^}]*)*\}'
        
        def replace_immini(match):
            # Extract first argument (content1)
            first_arg = match.group(1)
            return first_arg
        
        content = re.sub(immini_pattern, replace_immini, content)
        return content
    
    @classmethod
    def remove_image_commands(cls, content: str) -> str:
        """
        Step 4: Remove image commands and environments.
        
        Args:
            content: Content to clean
            
        Returns:
            Content with images removed
        """
        for pattern in cls.IMAGE_PATTERNS:
            content = re.sub(pattern, '', content, flags=re.DOTALL)
        return content
    
    @classmethod
    def remove_answer_commands(cls, content: str) -> str:
        """
        Step 5: Remove answer commands.
        
        Args:
            content: Content to clean
            
        Returns:
            Content with answer commands removed
        """
        # Use bracket-aware parsing for better accuracy
        for pattern in cls.ANSWER_PATTERNS:
            content = re.sub(pattern, '', content, flags=re.DOTALL)
        
        # Additional cleanup for remaining answer fragments
        content = re.sub(r'\{\\True[^}]*\}', '', content)
        content = re.sub(r'\{[^}]*\}(?=\s*\{)', '', content)  # Remove isolated answer options
        
        return content
    
    @classmethod
    def remove_solution_section(cls, content: str) -> str:
        """
        Step 6: Remove solution section (\loigiai{...}).
        
        Args:
            content: Content to clean
            
        Returns:
            Content with solution removed
        """
        # Use bracket-aware parsing for nested braces
        loigiai_start = content.find('\\loigiai')
        if loigiai_start != -1:
            # Find the opening brace after \loigiai
            brace_start = content.find('{', loigiai_start)
            if brace_start != -1:
                # Extract content using bracket parser
                solution_content = BracketParser.extract_content_from_braces(content, brace_start)
                # Remove the entire \loigiai{...} block
                solution_end = brace_start + len(solution_content) + 2  # +2 for braces
                content = content[:loigiai_start] + content[solution_end:]
        
        return content
    
    @classmethod
    def normalize_whitespace(cls, content: str) -> str:
        """
        Step 7: Normalize whitespace.
        
        Args:
            content: Content to normalize
            
        Returns:
            Content with normalized whitespace
        """
        # Remove leading/trailing whitespace
        content = content.strip()
        
        # Replace multiple spaces with single space
        content = re.sub(r' +', ' ', content)
        
        # Remove empty lines but preserve paragraph breaks
        lines = content.split('\n')
        cleaned_lines = []
        
        for line in lines:
            line = line.strip()
            if line:  # Keep non-empty lines
                cleaned_lines.append(line)
            elif cleaned_lines and cleaned_lines[-1]:  # Add single empty line for paragraph break
                cleaned_lines.append('')
        
        # Remove trailing empty lines
        while cleaned_lines and not cleaned_lines[-1]:
            cleaned_lines.pop()
        
        return '\n'.join(cleaned_lines)
    
    @classmethod
    def extract_clean_content(cls, latex_content: str) -> Tuple[str, str]:
        """
        Apply all 7 steps to extract clean content.
        
        Args:
            latex_content: Full LaTeX content
            
        Returns:
            Tuple of (raw_content, clean_content)
        """
        # Step 1: Extract from ex environment
        raw_content = cls.extract_from_ex_environment(latex_content)
        
        if not raw_content:
            return "", ""
        
        # Apply cleaning steps
        content = raw_content
        
        # Step 2: Remove metadata patterns
        content = cls.remove_metadata_patterns(content)
        
        # Step 3: Handle layout commands
        content = cls.handle_layout_commands(content)
        
        # Step 4: Remove image commands
        content = cls.remove_image_commands(content)
        
        # Step 5: Remove answer commands
        content = cls.remove_answer_commands(content)
        
        # Step 6: Remove solution section
        content = cls.remove_solution_section(content)
        
        # Step 7: Normalize whitespace
        content = cls.normalize_whitespace(content)
        
        return raw_content, content

File: apps/backend/internal/latex/content_extractor.go
package latex

import (
	"regexp"
	"strings"
)

// ContentExtractor implements 7-step content cleaning process
type ContentExtractor struct {
	bp *BracketParser
}

// NewContentExtractor creates a new content extractor
func NewContentExtractor() *ContentExtractor {
	return &ContentExtractor{
		bp: NewBracketParser(),
	}
}

// ExtractCleanContent applies all 7 steps to extract clean content
func (ce *ContentExtractor) ExtractCleanContent(latexContent string) (rawContent, cleanContent string) {
	// Step 1: Extract from ex environment
	rawContent = ce.ExtractFromExEnvironment(latexContent)
	if rawContent == "" {
		return "", ""
	}

	// Apply cleaning steps
	content := rawContent

	// Step 2: Remove \begin{ex} and \end{ex} tags from clean content
	content = ce.removeExTags(content)

	// Step 3: Remove metadata patterns
	content = ce.RemoveMetadataPatterns(content)

	// Step 4: Handle layout commands
	content = ce.HandleLayoutCommands(content)

	// Step 5: Remove image commands
	content = ce.RemoveImageCommands(content)

	// Step 6: Remove answer commands and their content completely
	content = ce.RemoveAnswerCommands(content)

	// Step 7: Remove solution section
	content = ce.RemoveSolutionSection(content)

	// Step 8: Normalize whitespace
	content = ce.NormalizeWhitespace(content)

	return rawContent, content
}

// Step 1: Extract content from \begin{ex}...\end{ex} environment
func (ce *ContentExtractor) ExtractFromExEnvironment(latexContent string) string {
	environments := ce.bp.ExtractEnvironmentContent(latexContent, "ex")
	if len(environments) > 0 {
		return environments[0] // Return first ex environment
	}
	return ""
}

// removeExTags removes \begin{ex} and \end{ex} tags
func (ce *ContentExtractor) removeExTags(content string) string {
	content = regexp.MustCompile(`\\begin\{ex\}`).ReplaceAllString(content, "")
	content = regexp.MustCompile(`\\end\{ex\}`).ReplaceAllString(content, "")
	return content
}

// Step 2: Remove metadata patterns (questionCode, source, subcount)
func (ce *ContentExtractor) RemoveMetadataPatterns(content string) string {
	patterns := []string{
		`%\s*\[\s*[0-9A-Z]{5,6}(?:-[0-9A-Z])?\s*\]\s*%?`, // QuestionCode
		`%\s*\[\s*Nguồn:?[^\]]+\s*\]\s*%?`,               // Source
		`\[\s*[A-Z]{2}\.\d+\s*\]`,                        // Subcount
	}

	for _, pattern := range patterns {
		re := regexp.MustCompile(pattern)
		content = re.ReplaceAllString(content, "")
	}

	return content
}

// Step 3: Handle layout commands (immini)
func (ce *ContentExtractor) HandleLayoutCommands(content string) string {
	// Handle \immini[optional]{content1}{content2} commands
	// Extract only the first argument (content1) and discard the second (usually image)

	maxIterations := 10
	iteration := 0

	for strings.Contains(content, "\\immini") && iteration < maxIterations {
		iteration++

		// Use regex first for simple cases
		imminiPattern := `\\immini(?:\[[^\]]*\])?\s*\{([^{}]*)\}\s*\{[^{}]*\}`
		re := regexp.MustCompile(imminiPattern)

		newContent := re.ReplaceAllString(content, "$1")

		// If regex worked, use the result
		if newContent != content {
			content = newContent
			continue
		}

		// Handle nested cases with bracket-aware parsing
		content = ce.handleNestedImmini(content)
		break
	}

	return content
}

// handleNestedImmini handles nested \immini commands with bracket-aware parsing
func (ce *ContentExtractor) handleNestedImmini(content string) string {
	for strings.Contains(content, "\\immini") {
		startPos := strings.Index(content, "\\immini")
		if startPos == -1 {
			break
		}

		// Find the start of the first brace
		firstBrace := strings.Index(content[startPos:], "{")
		if firstBrace == -1 {
			break
		}
		firstBrace += startPos

		// Extract first argument using bracket parser
		firstArg := ce.bp.ExtractContentFromBraces(content, firstBrace)
		if firstArg == "" {
			break
		}

		// Find end of first argument
		firstArgEnd := ce.findClosingBrace(content, firstBrace)
		if firstArgEnd == -1 {
			break
		}

		// Find second argument start
		secondBrace := strings.Index(content[firstArgEnd+1:], "{")
		if secondBrace == -1 {
			// No second argument, just replace with first
			content = content[:startPos] + firstArg + content[firstArgEnd+1:]
			continue
		}
		secondBrace += firstArgEnd + 1

		// Find end of second argument
		secondArgEnd := ce.findClosingBrace(content, secondBrace)
		if secondArgEnd == -1 {
			break
		}

		// Replace entire \immini command with just the first argument
		content = content[:startPos] + firstArg + content[secondArgEnd+1:]
	}

	return content
}

// Step 4: Remove image commands and environments
func (ce *ContentExtractor) RemoveImageCommands(content string) string {
	patterns := []string{
		`\\begin\{center\}[\s\S]*?\\end\{center\}`,
		`\\begin\{tikzpicture\}[\s\S]*?\\end\{tikzpicture\}`,
		`\\includegraphics(?:\[[^\]]*\])?\{[^}]*\}`,
		`\\begin\{figure\}[\s\S]*?\\end\{figure\}`,
	}

	for _, pattern := range patterns {
		re := regexp.MustCompile(pattern)
		content = re.ReplaceAllString(content, "")
	}

	return content
}

// Step 5: Remove answer commands and their content completely
func (ce *ContentExtractor) RemoveAnswerCommands(content string) string {
	// Protect LaTeX math expressions before processing
	protectedPatterns := ce.protectMathExpressions(content)

	// Find positions of answer commands and cut content before them
	answerPositions := ce.findAnswerCommandPositions(protectedPatterns.content)

	if len(answerPositions) > 0 {
		// Cut content at the first answer command position
		cutPosition := answerPositions[0]
		protectedPatterns.content = protectedPatterns.content[:cutPosition]
	}

	// Clean up remaining fragments
	protectedPatterns.content = ce.cleanAnswerFragments(protectedPatterns.content)

	// Restore protected LaTeX math patterns
	content = ce.restoreMathExpressions(protectedPatterns)

	return content
}

// protectedContent holds content with protected math expressions
type protectedContent struct {
	content      string
	replacements []struct {
		placeholder string
		original    string
	}
}

// protectMathExpressions protects LaTeX math expressions from being removed
func (ce *ContentExtractor) protectMathExpressions(content string) *protectedContent {
	pc := &protectedContent{content: content}

	// Protect \dfrac{}{} patterns
	dfracRe := regexp.MustCompile(`\\dfrac\{([^}]*)\}\{([^}]*)\}`)
	matches := dfracRe.FindAllStringSubmatch(content, -1)
	for i, match := range matches {
		placeholder := "__DFRAC_" + string(rune('A'+i)) + "__"
		pc.content = strings.Replace(pc.content, match[0], placeholder, 1)
		pc.replacements = append(pc.replacements, struct {
			placeholder string
			original    string
		}{placeholder, match[0]})
	}

	// Protect \frac{}{} patterns
	fracRe := regexp.MustCompile(`\\frac\{([^}]*)\}\{([^}]*)\}`)
	matches = fracRe.FindAllStringSubmatch(pc.content, -1)
	for i, match := range matches {
		placeholder := "__FRAC_" + string(rune('A'+i)) + "__"
		pc.content = strings.Replace(pc.content, match[0], placeholder, 1)
		pc.replacements = append(pc.replacements, struct {
			placeholder string
			original    string
		}{placeholder, match[0]})
	}

	// Protect other math commands
	mathCommands := []string{"sqrt", "sum", "int", "lim", "sin", "cos", "tan", "log", "ln"}
	for _, cmd := range mathCommands {
		pattern := `\\` + cmd + `\{([^}]*)\}`
		re := regexp.MustCompile(pattern)
		matches = re.FindAllStringSubmatch(pc.content, -1)
		for i, match := range matches {
			placeholder := "__" + strings.ToUpper(cmd) + "_" + string(rune('A'+i)) + "__"
			pc.content = strings.Replace(pc.content, match[0], placeholder, 1)
			pc.replacements = append(pc.replacements, struct {
				placeholder string
				original    string
			}{placeholder, match[0]})
		}
	}

	return pc
}

// findAnswerCommandPositions finds positions of answer commands
func (ce *ContentExtractor) findAnswerCommandPositions(content string) []int {
	var positions []int

	// Find \choiceTF
	if pos := strings.Index(content, "\\choiceTF"); pos != -1 {
		positions = append(positions, pos)
	}

	// Find \choice (but not \choiceTF)
	pos := 0
	for {
		choicePos := strings.Index(content[pos:], "\\choice")
		if choicePos == -1 {
			break
		}
		choicePos += pos

		// Check if it's not \choiceTF
		if choicePos+9 >= len(content) || content[choicePos:choicePos+9] != "\\choiceTF" {
			positions = append(positions, choicePos)
			break
		}
		pos = choicePos + 1
	}

	// Find \shortans
	if pos := strings.Index(content, "\\shortans"); pos != -1 {
		positions = append(positions, pos)
	}

	// Find \matching
	if pos := strings.Index(content, "\\matching"); pos != -1 {
		positions = append(positions, pos)
	}

	return positions
}

// cleanAnswerFragments removes remaining answer fragments
func (ce *ContentExtractor) cleanAnswerFragments(content string) string {
	// Remove True/False markers
	content = regexp.MustCompile(`\{\\True\s+[^}]*\}`).ReplaceAllString(content, "")
	content = regexp.MustCompile(`\{\\False\s+[^}]*\}`).ReplaceAllString(content, "")
	content = regexp.MustCompile(`\\True\s*`).ReplaceAllString(content, "")
	content = regexp.MustCompile(`\\False\s*`).ReplaceAllString(content, "")

	return content
}

// restoreMathExpressions restores protected math expressions
func (ce *ContentExtractor) restoreMathExpressions(pc *protectedContent) string {
	content := pc.content
	for _, replacement := range pc.replacements {
		content = strings.Replace(content, replacement.placeholder, replacement.original, 1)
	}
	return content
}

// Step 6: Remove solution section (\loigiai{...})
func (ce *ContentExtractor) RemoveSolutionSection(content string) string {
	for {
		loigiaiStart := strings.Index(content, "\\loigiai")
		if loigiaiStart == -1 {
			break
		}

		// Find the opening brace after \loigiai
		braceStart := strings.Index(content[loigiaiStart:], "{")
		if braceStart == -1 {
			break
		}
		braceStart += loigiaiStart

		// Find the matching closing brace
		solutionEnd := ce.findMatchingSolutionEnd(content, braceStart)
		if solutionEnd == -1 {
			// No matching brace found, remove from loigiai to end
			content = content[:loigiaiStart]
			break
		}

		// Remove the entire solution section
		content = content[:loigiaiStart] + content[solutionEnd+1:]
	}

	return content
}

// findMatchingSolutionEnd finds the matching closing brace for solution
func (ce *ContentExtractor) findMatchingSolutionEnd(content string, braceStart int) int {
	braceCount := 0
	pos := braceStart
	maxSearchLength := len(content) - braceStart
	searchCount := 0

	for pos < len(content) {
		// Safety check to prevent infinite loops
		searchCount++
		if searchCount > maxSearchLength {
			return -1
		}

		char := content[pos]

		// Handle escaped characters
		if char == '\\' && pos+1 < len(content) {
			pos += 2
			continue
		}

		if char == '{' {
			braceCount++
		} else if char == '}' {
			braceCount--
			if braceCount == 0 {
				return pos
			}
		}

		pos++
	}

	return -1 // No matching brace found
}

// Step 7: Normalize whitespace
func (ce *ContentExtractor) NormalizeWhitespace(content string) string {
	// Convert newlines to spaces for better text flow
	content = strings.ReplaceAll(content, "\r\n", " ")
	content = strings.ReplaceAll(content, "\r", " ")
	content = strings.ReplaceAll(content, "\n", " ")

	// Replace multiple spaces with single space
	content = regexp.MustCompile(`[ ]+`).ReplaceAllString(content, " ")

	// Replace multiple tabs with spaces
	content = regexp.MustCompile(`\t+`).ReplaceAllString(content, " ")

	// Trim leading and trailing whitespace
	content = strings.TrimSpace(content)

	return content
}

// findClosingBrace finds the position of closing brace with bracket counting
func (ce *ContentExtractor) findClosingBrace(content string, startPos int) int {
	if startPos >= len(content) || content[startPos] != '{' {
		return -1
	}

	braceLevel := 1
	pos := startPos + 1

	for pos < len(content) && braceLevel > 0 {
		char := content[pos]

		// Handle escaped characters
		if char == '\\' && pos+1 < len(content) {
			pos += 2
			continue
		}

		if char == '{' {
			braceLevel++
		} else if char == '}' {
			braceLevel--
		}

		pos++
	}

	if braceLevel == 0 {
		return pos - 1 // Position of closing brace
	}

	return -1
}
